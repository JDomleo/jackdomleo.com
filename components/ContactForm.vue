<template>
  <form method="POST" name="contact" data-netlify="true" netlify-honeypot="bot-field">
    <div class="contact">
      <input type="hidden" name="form-name" value="contact">
      <div style="display: none">
        <label for="bot-field">
          Do not fill out this field if you are human!
        </label>
        <input id="bot-field" name="bot-field">
      </div>
      <div class="textfield">
        <label for="contact-name">
          Name<span aria-hidden="true">
            *
          </span> <span class="sr-only">
            (required)
          </span>
        </label>
        <span class="textfield__entry">
          <svg-icon name="user" />
          <input
            id="contact-name"
            v-model="contactForm.name"
            name="name"
            type="text"
            inputmode="text"
            autocomplete="name"
            placeholder="Mickey Mouse"
            required
            :aria-invalid="$v.contactForm.name.$dirty && $v.contactForm.name.$invalid.toString()"
            :aria-describedby="$v.contactForm.name.$dirty && $v.contactForm.name.$invalid && 'contact-name--error'"
            @change="autosave"
            @input="touch('name')"
            @blur="touch('name')"
          >
        </span>
        <span v-if="$v.contactForm.name.$error" id="contact-name--error" class="textfield__error" role="alert">
          This field is required
        </span>
      </div>
      <div class="textfield">
        <label for="contact-email">
          Email<span aria-hidden="true">
            *
          </span> <span class="sr-only">
            (required)
          </span>
        </label>
        <span class="textfield__entry">
          <svg-icon name="email" />
          <input
            id="contact-email"
            v-model="contactForm.email"
            name="email"
            type="email"
            inputmode="email"
            autocomplete="email"
            placeholder="mickey.mouse@example.com"
            required
            :aria-invalid="$v.contactForm.email.$dirty && $v.contactForm.email.$invalid.toString()"
            :aria-describedby="$v.contactForm.email.$dirty && $v.contactForm.email.$invalid && 'contact-email--error'"
            @change="autosave"
            @input="touch('email')"
            @blur="touch('email')"
          >
        </span>
        <span v-if="$v.contactForm.email.$error" id="contact-email--error" class="textfield__error" role="alert">
          <template v-if="!$v.contactForm.email.email">
            Enter a valid email address
          </template><template v-else-if="!$v.contactForm.email.required">
            This field is required
          </template>
        </span>
      </div>
      <div class="textfield">
        <label for="contact-message">
          Message<span aria-hidden="true">
            *
          </span> <span class="sr-only">
            (required)
          </span>
        </label>
        <span class="textfield__entry">
          <svg-icon name="pencil" />
          <textarea
            id="contact-message"
            v-model="contactForm.message"
            name="message"
            type="text"
            inputmode="text"
            placeholder="Jack, I am contacting you on this fine day because..."
            required
            :aria-invalid="$v.contactForm.message.$dirty && $v.contactForm.message.$invalid.toString()"
            :aria-describedby="$v.contactForm.message.$dirty && $v.contactForm.message.$invalid && 'contact-message--error'"
            @change="autosave"
            @input="touch('message')"
            @blur="touch('message')"
          />
        </span>
        <span v-if="$v.contactForm.message.$error" id="contact-message--error" class="textfield__error" role="alert">
          This field is required
        </span>
      </div>
      <button class="btn" type="submit" :disabled="$v.contactForm.$invalid" :aria-label="$v.contactForm.$invalid ? 'Some required fields are missing' : undefined" :data-cooltipz-dir="$v.contactForm.$invalid ? 'left' : undefined">
        Send
      </button>
    </div>
  </form>
</template>

<script lang="ts">
import { Component, Vue } from 'nuxt-property-decorator'

import { email, required } from 'vuelidate/lib/validators'
import { validationMixin } from 'vuelidate'

interface IContactForm {
  name: string;
  email: string;
  message: string;
}

@Component({
  mixins: [validationMixin],
  validations: {
    contactForm: {
      name: {
        required
      },
      email: {
        email,
        required
      },
      message: {
        required
      }
    }
  }
})
export default class ContactForm extends Vue {
  private contactForm: IContactForm = {
    name: '',
    email: '',
    message: ''
  };

  private mounted (): void {
    const data = sessionStorage.getItem('autosave')

    if (data) {
      this.contactForm = JSON.parse(data)
    }
  }

  private autosave (): void {
    sessionStorage.setItem('autosave', JSON.stringify(this.contactForm))
  }

  private touch (formField: string): void {
    return this.$v.contactForm[formField]!.$touch()
  }
}
</script>

<style lang="scss">
:root {
  --contact-border: var(--body-color);
  --contact-placeholder: var(--base-grey-middle);
}

.theme--dark {
  --contact-placeholder: var(--base-grey-lighter);
}
</style>

<style lang="scss" scoped>
.contact {
  display: flex;
  flex-direction: column;
  gap: 1rem;

  button[type="submit"] {
    align-self: flex-end;
  }
}

.textfield {
  display: flex;
  flex-direction: column;
  gap: 0.2rem;

  label {
    font-size: 1.2rem;

    span[aria-hidden="true"] {
      color: var(--alt-orange);
      user-select: none;
    }
  }

  &__entry {
    display: flex;
    gap: 0.5rem;
    border: 2px solid var(--contact-border);
    padding: 0.5rem;
    border-radius: 0.5rem;

    &:focus-within {
      border-color: var(--alt-orange);

      svg {
        color: var(--alt-orange);
      }
    }

    svg {
      height: 2.5rem;
      width: 2.5rem;
      padding-right: 0.5rem;
      border-right: 2px solid currentColor;
    }

    input,
    textarea {
      flex: 1;
      background-color: transparent;
      color: currentColor;
      border: none;

      &::placeholder {
        color: var(--contact-placeholder);
      }

      &:focus {
        outline: none;
      }
    }

    textarea {
      resize: vertical;
      min-height: 25vmin;

      @media (min-width: 550px) {
        min-height: 15vmin;
      }
    }
  }

  &__error {
    color: var(--base-red);
    font-weight: 700;
  }
}

.btn {
  border: 2px solid currentColor;
  background-color: transparent;
  border-radius: 0.5rem;
  padding: 0.8rem 1.6rem;
  cursor: pointer;
  transition: background-color 160ms ease;

  &:not( [disabled] ) {
    &:hover,
    &:focus {
      background-color: var(--base-orange);
    }
  }

  &:disabled {
    opacity: 0.4;
    cursor: default;
  }
}
</style>
