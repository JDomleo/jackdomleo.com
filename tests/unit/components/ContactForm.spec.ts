import { mount, shallowMount } from '@vue/test-utils'
import { configureAxe, toHaveNoViolations } from 'jest-axe'

import ContactForm from '@/components/ContactForm.vue'

expect.extend(toHaveNoViolations)

const axe = configureAxe({
  rules: {
    region: { enabled: false } // Don't need to check because this component is isolated
  }
})

describe('ContactForm.vue component', () => {
  /**
   * Vue instance
   */

  it('is a Vue instance', () => {
    const wrapper = mount(ContactForm, {
      stubs: ['svg-icon']
    })

    expect(wrapper.vm).toBeTruthy()
  })

  /**
   * Snapshots
   */

  it('matches snapshot when mounted', () => {
    const wrapper = mount(ContactForm, {
      stubs: ['svg-icon']
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('matches snapshot when shallow mounted', () => {
    const wrapper = shallowMount(ContactForm, {
      stubs: ['FormTextInput']
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  /**
   * A11y
   */

  it('to have no obvious accessibility violations', async () => {
    const wrapper = mount(ContactForm, {
      stubs: ['svg-icon']
    })

    expect(await axe(wrapper.element)).toHaveNoViolations()
  })

  /**
   * Interactivity
   */

  it('enables the send button when all fields have been filled in', async () => {
    const wrapper = mount(ContactForm, {
      stubs: ['svg-icon']
    })

    const sendButton = wrapper.find('button[type="submit"]')

    expect(sendButton.attributes('disabled')).toBe('disabled')

    await wrapper.find('input#contact-name').setValue('Mickey Mouse')
    await wrapper.find('input#contact-email').setValue('mickey.mouse@example.com')
    await wrapper.find('textarea#contact-message').setValue('Hello Jack!')

    expect(sendButton.attributes('disabled')).toBeUndefined()

    await wrapper.find('input#contact-name').setValue('')
    await wrapper.find('input#contact-email').setValue('')
    await wrapper.find('textarea#contact-message').setValue('')

    expect(sendButton.attributes('disabled')).toBe('disabled')
  })

  it('only shows errors if the form field is dirty and is invalid', async () => {
    const wrapper = mount(ContactForm, {
      stubs: ['svg-icon']
    })

    expect(wrapper.find('span#contact-name--error').exists()).toBe(false)
    expect(wrapper.find('span#contact-email--error').exists()).toBe(false)
    expect(wrapper.find('span#contact-message--error').exists()).toBe(false)

    await wrapper.find('input#contact-name').setValue('Mickey Mouse')
    await wrapper.find('input#contact-email').setValue('mickey.mouse@example.com')
    await wrapper.find('textarea#contact-message').setValue('Hello Jack!')

    expect(wrapper.find('span#contact-name--error').exists()).toBe(false)
    expect(wrapper.find('span#contact-email--error').exists()).toBe(false)
    expect(wrapper.find('span#contact-message--error').exists()).toBe(false)

    await wrapper.find('input#contact-name').setValue('')
    await wrapper.find('input#contact-email').setValue('')
    await wrapper.find('textarea#contact-message').setValue('')

    expect(wrapper.find('span#contact-name--error').exists()).toBe(true)
    expect(wrapper.find('span#contact-email--error').exists()).toBe(true)
    expect(wrapper.find('span#contact-message--error').exists()).toBe(true)

    await wrapper.find('input#contact-name').setValue('Mickey Mouse')
    await wrapper.find('input#contact-email').setValue('mickey.mouse@example.com')
    await wrapper.find('textarea#contact-message').setValue('Hello Jack!')

    expect(wrapper.find('span#contact-name--error').exists()).toBe(false)
    expect(wrapper.find('span#contact-email--error').exists()).toBe(false)
    expect(wrapper.find('span#contact-message--error').exists()).toBe(false)
  })
})
