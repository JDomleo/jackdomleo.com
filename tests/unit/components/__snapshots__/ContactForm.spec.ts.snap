// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm.vue component matches snapshot when mounted 1`] = `
<form
  data-netlify="true"
  method="POST"
  name="contact"
  netlify-honeypot="bot-field"
>
  <div
    class="contact"
  >
    <input
      name="form-name"
      type="hidden"
      value="contact"
    />
     
    <div
      style="display: none;"
    >
      <label
        for="bot-field"
      >
        
        Do not fill out this field if you are human!
      
      </label>
       
      <input
        id="bot-field"
        name="bot-field"
      />
    </div>
     
    <div
      class="text-input"
    >
      <label
        for="contact-name"
      >
        
    Name
    
        <span
          aria-hidden="true"
        >
          
        *
      
        </span>
         
        <span
          class="sr-only"
        >
          
        (required)
      
        </span>
      </label>
       
      <span
        class="text-input__entry"
      >
        <svg-icon-stub
          name="user"
        />
         
        <input
          aria-invalid="false"
          autocomplete="name"
          id="contact-name"
          inputmode="text"
          name="name"
          placeholder="Mickey Mouse"
          required="required"
          type="text"
        />
      </span>
       
      <!---->
    </div>
     
    <div
      class="text-input"
    >
      <label
        for="contact-email"
      >
        
    Email
    
        <span
          aria-hidden="true"
        >
          
        *
      
        </span>
         
        <span
          class="sr-only"
        >
          
        (required)
      
        </span>
      </label>
       
      <span
        class="text-input__entry"
      >
        <svg-icon-stub
          name="email"
        />
         
        <input
          aria-invalid="false"
          autocomplete="email"
          id="contact-email"
          inputmode="email"
          name="email"
          placeholder="mickey.mouse@example.com"
          required="required"
          type="email"
        />
      </span>
       
      <!---->
    </div>
     
    <div
      class="text-input"
    >
      <label
        for="contact-message"
      >
        
    Message
    
        <span
          aria-hidden="true"
        >
          
        *
      
        </span>
         
        <span
          class="sr-only"
        >
          
        (required)
      
        </span>
      </label>
       
      <span
        class="text-input__entry"
      >
        <svg-icon-stub
          name="pencil"
        />
         
        <textarea
          aria-invalid="false"
          id="contact-message"
          inputmode="text"
          name="message"
          placeholder="Greetings Jack! I hope you are doing swell..."
          required="required"
          type="text"
        />
      </span>
       
      <!---->
    </div>
     
    <button
      aria-label="Some required fields are missing"
      class="btn"
      data-cooltipz-dir="left"
      disabled="disabled"
      type="submit"
    >
      
      Send
    
    </button>
  </div>
</form>
`;

exports[`ContactForm.vue component matches snapshot when shallow mounted 1`] = `
<form
  data-netlify="true"
  method="POST"
  name="contact"
  netlify-honeypot="bot-field"
>
  <div
    class="contact"
  >
    <input
      name="form-name"
      type="hidden"
      value="contact"
    />
     
    <div
      style="display: none;"
    >
      <label
        for="bot-field"
      >
        
        Do not fill out this field if you are human!
      
      </label>
       
      <input
        id="bot-field"
        name="bot-field"
      />
    </div>
     
    <formtextinput-stub
      autocomplete="name"
      forid="contact-name"
      icon="user"
      inputmode="text"
      label="Name"
      name="name"
      placeholder="Mickey Mouse"
      required="true"
      type="text"
      value=""
    />
     
    <formtextinput-stub
      autocomplete="email"
      forid="contact-email"
      icon="email"
      inputmode="email"
      label="Email"
      name="email"
      placeholder="mickey.mouse@example.com"
      required="true"
      type="email"
      value=""
    />
     
    <formtextinput-stub
      forid="contact-message"
      icon="pencil"
      inputmode="text"
      label="Message"
      multiline="true"
      name="message"
      placeholder="Greetings Jack! I hope you are doing swell..."
      required="true"
      type="text"
      value=""
    />
     
    <button
      aria-label="Some required fields are missing"
      class="btn"
      data-cooltipz-dir="left"
      disabled="disabled"
      type="submit"
    >
      
      Send
    
    </button>
  </div>
</form>
`;
